### PR TITLE
Honor the users privacy decision and do not send UDP packets if he does not want to send metrics

### PR DESCRIPTION
**Pull request by clemensg**
Hi, if the user does not want to send metrics and answers N to your question, you should honor that decision.
Instead what you are doing is, you still send metrics, but replace the fields in the hash with "anonymous".
Let me tell you: I don't want to send metrics, not even metrics with fields filled with "anonymous".
Not wanting to send metrics does NOT mean: Please collect my IP address, time and date whenever I deploy something, it means: Do not send anything to the metrics server!
I corrected the collect function. It now exits if the user did not want to send metrics.
Regards, Clemens

### CODE REVIEW

**Comment by leehambley** [comment]
Hi @clemensg thanks for the PR. As you can see from [the implementation](https://github.com/capistrano/stats/tree/master/server), specifically [the storage format](https://github.com/capistrano/stats/blob/a5b26825a64eb3ca0b4e69a72d61a7f469a10fbd/server/server.go#L19), we don't store any IP addresses or any personal information if the user has elected not to track statistics. Indeed currently the server does not handle this case at all, panicking [here](https://github.com/capistrano/stats/blob/master/server/server.go#L33). 

We consider this to be the equivalent of an HTTP ping, with blocked cookies, an echo, ghost or empty impression which has no value except in a broad statistical sense.

I'm relieved to see someone performing a diligent code review on their third party dependencies.

We felt it was important to strike a balance between collecting meaningful numbers (that means being able to approximate number of downloads, compared with number of opted-in users, compared with number of opted-out users) in the short term to be able to gauge how many people _typically_ opt in to tracking given a choice. 

We publicise both the collector-client and the server so that people can make informed decisions, and as the code is so simple, it's relatively approachable for review.

Using all the standard tools of Rake, Capistrano and Ruby, if someone feels particularly strongly they can easily disable it be sticking with a pre-`3.3.3` version of Capistrano (less than ideal, I know), back-porting upstream changes into a custom fork can be a nightmare.

Preferably, in your case, if this issue hasn't undermined your trust in Capistrano, you might choose to do something like the following:

``` ruby
# Original definition https://github.com/capistrano/capistrano/blob/master/lib/capistrano/tasks/deploy.rake#L3
Rake::Task['deploy:starting'].clear_actions
namespace :deploy do
  task :starting do
    invoke 'deploy:check'
    invoke 'deploy:set_previous_revision'
  end
end
```

Perhaps simpler still:

``` ruby
Rake::Task['metrics:collect'].clear_actions
```

YMMV, neither of these snippets are tested.

I hope for the benefit of the project you do choose to enable statistics, and barring that, I will broaden in the stead of this ticket the documentation in both the `stats` and `capistrano` repositories to make clear that _opt-out_ does not mean _absolutely nothing sent_.

I'll leave this ticket open for now as a point of discussion. I can't stress how difficult it is as a maintainer of something as broadly useful as Capistrano to maintain it in a logical way without knowing what fraction of users are using a particular combination of platform, interpreter version, etc; and I sincerely need the data in order to make determinations about whether the data I receive from the opted-in users is a representative sample or not. 

And, naturally the project anon hash cannot be turned into anything of an identifying nature, and this is used solely to determine whether many people deploy many project infrequently, or few people deploy few projects very frequently.

Soon the data will be public anyway, via a JSON API and a website which will periodically make the entire database available for download.

**Comment by clemensg** [comment]
Hi, thanks for taking my concerns seriously and thanks for your workaround suggestions.

I do believe you that you do not collect that data for the moment, however: We can't be sure if the code on the repo will always be the same as the code running on the metrics server.
On my private machine and for open source projects, I might even enable the stats to help the project, but I do not like the idea that suddenly you have to ping the metrics server even if you opted out.

Of course I can hack around it, but most people won't even notice that there is a connection going out. They will assume, that if they select No, you do not send metrics, not even one UDP packet to the metrics server.

What do you think about asking the user if he agrees to this UDP ping in case he opted out of the metrics?

Another thing bothered me: The exception which is thrown when connections to the metric server are blocked by network filters like LittleSnitch or in case the Internet connection is down and the UDP packet can't be sent to the metrics server. Still I might use Capistrano within my LAN and don't want that deployment to fail, just because the metrics server can't be reached.

```
(Backtrace restricted to imported tasks)
cap aborted!
Errno::EHOSTDOWN: Host is down - sendto(2) for "metrics.capistranorb.com" port 1200

Tasks: TOP => metrics:collect
(See full trace by running task with --trace)
The deploy has failed with an error: Host is down - sendto(2) for "metrics.capistranorb.com" port 1200
```

Metrics should not be a dependency for the program to work correctly, right?

By the way: `Rake::Task['metrics:collect'].clear_actions` works fine! Thank you.

**Comment by leehambley** [comment]
Thanks for the quick reply @clemensg, I've quoted some parts of your message and will reply inline.

> I do believe you that you do not collect that data for the moment, however: We can't be sure if the code on the repo will always be the same as the code running on the metrics server.

That is a valid problem, and something I sincerely wish we had a solution for (for many, many cases). Until signing binaries, and something kind of HMAC/signature pinning between client and server becomes easy to implement, and reliable, I'm not sure you do have a way to verify that the code in this repository is what we are running on the collector side, and that makes me (in a general sense as an open-source citizen) as uncomfortable as it makes you.

It did occur to me that there might be something I could do with a challenge/response with secret based on the current git revision, but in the end, it still solved the wrong problem.

I notice you are from a similar background technology-wise as I am, and perhaps we could get our heads together and figure out a solution, but (see below) indeed that might _also_ be solving the wrong problem.

> Of course I can hack around it, but most people won't even notice that there is a connection going out. They will assume, that if they select No, you do not send metrics, not even one UDP packet to the metrics server.

I'm comfortable with that, many people also have no idea how much data is leaked every time they open an app, browse a website, open their phone, use location services, etc. Even how much data is leaked by any of their devices and software intentionally, or unintentionally, I feel that we are _firmly_ on the white-hat side of that discussion, since we have placed all the implementation code in the public domain.

> What do you think about asking the user if he agrees to this UDP ping in case he opted out of the metrics?

That sounds ideal, I might have implemented it that way myself, except for a lack of time, the 3.3.3 release of Capistrano contained more than 85 patches from various people in the end, and my wife recently had a baby, and we're going through the motions of buying a house, as one-man-teams go, I was a little overloaded. (And overwhelmed by discussions about platform support for various tweaks people wanted to make for Cap, thus driving the metrics implementation.)

> Another thing bothered me: The exception which is thrown when connections to the metric server are blocked by network filters like LittleSnitch or in case the metric server is down.
> `...`
> Metrics should not be a dependency for the program to work correctly.
> Just assume your metrics server is down, you don't want every deployment to fail, right?

Absolutely not, in fact the reason I chose a UDP packet was to avoid that, in my testing I only noticed that when the DNS is not routable a `SocketError` could be raised, which I [recently patched for](https://github.com/capistrano/stats/blame/e6724619694ae4a1cf5f74d33c458377e1236b7e/gem/lib/capistrano-stats/metric-collector.rb#L24). I confess however not to having blocked the port with a firewall or similar, as I naïvely considered that it couldn't fail.

---

To summarise, if you would be willing (although it is by far your least preferred solution, I am sure) to make a PR to implement the "don't even ping" solution, I would be more than happy to merge and push a new release almost immediately (after reviewing), we'd also have to find a way to encode that meaning into the so-far boolean `./.capistranometrics` file.

If you can't, or don't want to take the time, I'm sure I'll eventually get around to it, but I barely had time to write _this_ version, and am again being pulled in rather too many directions.

Thanks again for the dialog, I look forward to reviewing some code with you, if you can.

**Comment by clemensg** [comment]
That's a difficult problem indeed, I will have to think about it.

Btw. I created a PR with a first attempt to fix the mentioned problems.

**Comment by leehambley** [comment]
> Btw. I created a PR with a first attempt to fix the mentioned problems.

Thanks, Github already linked since you commented, but incase anyone is following this thread by email, see #2.
